diff --git a/src/ch.usi.inf.nodeprof/js/jalangi.js b/src/ch.usi.inf.nodeprof/js/jalangi.js
index ae93453..3f7754f 100644
--- a/src/ch.usi.inf.nodeprof/js/jalangi.js
+++ b/src/ch.usi.inf.nodeprof/js/jalangi.js
@@ -155,21 +155,36 @@ process.on('SIGINT', function(){
 process.on('exit', function () { J$.endExecution(); });
 
 path=require('path');
-while(process.argv.length > 3) {
-  var toLoad = process.argv[2];
-  try{
-    require(path.resolve(toLoad));
-  }catch(e1){
-    console.log("error loading "+process.argv[2]);
-    console.trace(e1);
-    process.exit(-1);
-  }
-  process.argv = process.argv.slice(1);
+if (process.argv.length < 4) {
+  console.log("At least 4 arguments required");
+  process.exit(-1);
+}
+
+try {
+  require(path.resolve(process.argv[2]));
+} catch(err) {
+  console.log("error loading " + process.argv[2]);
+  console.trace(err);
+  process.exit(-1);
 }
-if(process.argv.length == 3) {
-  process.argv = process.argv.slice(1);
+
+process.argv.splice(1,2);
+process.argv[0] = process.argv0
+process.execPath = process.argv0
+
+if (process.argv[1] == '-e' || process.argv[1] == '--eval') {
+  (function() {
+    var script = 'global.__filename = "[eval]";\n' +
+                 'global.exports = exports;\n' +
+                 'global.module = module;\n' +
+                 'global.__dirname = __dirname;\n' +
+                 'global.require = require;\n' +
+                 'return require("vm").runInThisContext(' +
+                 JSON.stringify(process.argv[2]) +
+                 ')';
+    module._compile(script, '[eval]-wrapper');
+  })();
+} else {
   process.argv[1] = path.resolve(process.argv[1]);
   require('module').runMain();
-}else {
-  console.log("[pathtoanalysis]* pathto.js");
 }
